### DISCUSSION TITLE
Import external dataset as CSV file

### DISCUSSION THREAD

**Discussion by rudilien** [idaholab/moose#21941] [Q&A General] [answered]
I would like to import temperature data into a MOOSE model. I attempted to test this by importing a CSV file with made-up temperature data corresponding to X and Y coordinates that should result in a small region (rectangular in shape) of elevated temperature at the center of the geometry. Instead, a rhombus-like shape of elevated temperature is displayed, as well as elevated temperatures on the left and right boundaries. Can I get help with how to fix this issue? I am relatively new to using MOOSE so detailed answers would be appreciated in responses! I've attached my input file (attached as .txt), CSV file to be imported (column 1: X coord; column 2: Y coord; column 3: Temp (K)), and what is currently displaying as the imported temperature data.

I would also like to know how to save either node ID or coordinates and MOOSE results (such as displacement and stress) as a CSV file.

Files:
[tensmech_DataImportTest.txt](https://github.com/idaholab/moose/files/9428589/tensmech_DataImportTest.txt)

[ExData_nodes.csv](https://github.com/idaholab/moose/files/9428593/ExData_nodes.csv)

<img width="468" alt="Test_tempimport" src="https://user-images.githubusercontent.com/98783914/186781146-f7cd4ddf-733e-4583-b84e-ad8accb2aac9.png">

**Comment by GiudGiud** [comment, marked as answer]
Hello

You are indeed seeing this because of numbering issues. The data numbering does not match your mesh ordering. Is this mesh from one of our mesh generators? Or from a file you imported?

how are you loading the CSV data? Which object and which options?

you should look at the vectorpostprocessor system or an auxkernel (elementIDaux andnodeidaux) to save/output the element/node numbering

Guillaume

**Reply by rudilien** [reply]
The mesh is made using the "GeneratedMeshGenerator", not an imported file. In future work, I will likely use an imported mesh.

I am loading the CSV data using "PropertyReadFile" with "read_type = 'node'" as a User Object and then the "PiecewiseConstantFromCSV" Function, also with "read_type = 'node'". To apply the data as an initial condition, I also use "FunctionIC" in the ICs block, which is linked to the function defined within the Functions block. I am attempting to import and apply the CSV data to nodes by specifying the XY coordinates of the nodes in the CSV file. It was unclear from the Moose website whether the CSV data should include node numbers or coordinates of the nodes when using "read_type = node", so maybe I need to use a different read_type setting to import data corresponding to coordinates? The input file that I made is attached in my original post as "tensmech_DataImportTest.txt" - you can look at this to see how I loaded the CSV data if my answer here is unclear.

I tried to follow data import methods that were described in other similar discussion posts in the Moose GitHub. I would like to know how to import data using XYZ (and eventually time) coordinates rather than element/node numbers so that I can import external data that may not have the same node numbering in the future.

**Reply by GiudGiud** [reply]
with the read_type = node the data should be ordered by node ID not by node coordinate.

to use coordinates you may use the read_type = grain or VORONOI. This could be a little slower as it triggers searching for the nearest point to find the appropriate data

**Reply by rudilien** [reply]
Thanks for the clarification on node vs. voronoi read_type. I have switched my User Object and Function blocks to have "read_type = voronoi" but now receive this warning when running the input:
_The following warning occurred in the object "reader_nearest", of type "PropertyReadFile".
Data size in /Users/rudilien/projects/moose/modules/tensor_mechanics/tutorials/introduction/MyScripts/ExData_nodes.csv is larger than 3 objects, some data will not be used._

The CSV file that I am importing has three columns: the first two are the X and Y coordinates, the third is the temperature data (see CSV attachment in my original post). I also recreated the CSV file in case something had been saved incorrectly. Is this related to the "nvoronoi" parameter? Could you provide more clarification on what the nvoronoi parameter should be set to as well? 

In addition to the warning, the temperature import does not include the elevated temperature in the center of the mesh - screenshot attached:
<img width="954" alt="Screen Shot 2022-08-30 at 11 48 21 AM" src="https://user-images.githubusercontent.com/98783914/187519907-7ea12619-e95a-46b2-9f4f-603755ebfdcf.png">

Thanks for your assistance!

**Reply by GiudGiud** [reply]
nvoronoi should be set to the number of nodes if the number of points you want to use is the same as the number of nodes

it's the number of nearest-neighbor regions

**Reply by rudilien** [reply]
Setting nvoronoi equal to the number of nodes fixed my issue. Thanks so much!
